backends/occa/ceed-occa-vector.c

```diff
@@ -31,187 +24,154 @@ static inline size_t bytes(const CeedVector vec) {
-  //occaSyncD2H(vec); // not good for t05
   return 0;
 }
-
-// *****************************************************************************
-// * Restore
 // *****************************************************************************
-static int CeedVectorRestoreArray_Occa(const CeedVector vec,
+static int CeedVectorRestoreArray_Occa(const CeedVector x,
                                        CeedScalar **array) {
-  //CeedVector_Occa *impl = vec->data;
-
-  CeedDebug("\033[33m[CeedVector][RestoreArray]");
-  //impl->array = NULL;
-  *array = NULL;
-  //occaSyncD2H(vec); // not good for t05
-  return 0;
+  return CeedVectorRestoreArrayRead_Occa(x,(const CeedScalar**)array);
 }
 
 // *****************************************************************************
-// * Destroy
+// * Destroy the vector
 // *****************************************************************************
 static int CeedVectorDestroy_Occa(const CeedVector vec) {
-  CeedVector_Occa *impl = vec->data;
   int ierr;
-
+  CeedVector_Occa *occa = vec->data;
   CeedDebug("\033[33m[CeedVector][Destroy]");
-  ierr = CeedFree(&impl->array_allocated); CeedChk(ierr);
-  ierr = CeedFree(&vec->data); CeedChk(ierr);
-  // free device memory
-  //occaMemoryFree(*impl->device);
-  // free device object
-  //CeedChk(CeedFree(&impl->device));
-  // free our CeedVector_Occa struct
+  occaMemoryFree(*occa->d_array);
+  ierr = CeedFree(&occa->h_array); CeedChk(ierr);
+  ierr = CeedFree(&occa->d_array); CeedChk(ierr);
+  ierr = CeedFree(&occa); CeedChk(ierr);
   return 0;
 }
 
 // *****************************************************************************
-// * Create
+// * Create a vector of the specified length (does not allocate memory)
 // *****************************************************************************
 int CeedVectorCreate_Occa(const Ceed ceed, const CeedInt n, CeedVector vec) {
-  CeedVector_Occa *impl;
-  Ceed_Occa *ceed_data=ceed->data;
   int ierr;
-
+  CeedVector_Occa *data;
+  const Ceed_Occa *occa=ceed->data;
   CeedDebug("\033[33m[CeedVector][Create] n=%d", n);
-  // ***************************************************************************
   vec->SetArray = CeedVectorSetArray_Occa;
   vec->GetArray = CeedVectorGetArray_Occa;
   vec->GetArrayRead = CeedVectorGetArrayRead_Occa;
   vec->RestoreArray = CeedVectorRestoreArray_Occa;
   vec->RestoreArrayRead = CeedVectorRestoreArrayRead_Occa;
   vec->Destroy = CeedVectorDestroy_Occa;
-  // Allocating impl, host & device
-  ierr = CeedCalloc(1,&impl); CeedChk(ierr);
-  ierr = CeedCalloc(1,&impl->device); CeedChk(ierr);
-  *impl->device = occaDeviceMalloc(ceed_data->device, bytes(vec), NULL, NO_PROPS);
-  vec->data = impl;
+  // ***************************************************************************
+  ierr = CeedCalloc(1,&data); CeedChk(ierr);
+  vec->data = data;
+  // ***************************************************************************
+  ierr = CeedCalloc(1,&data->d_array); CeedChk(ierr);
```

Comment:
I don't understand this indirection. If there is only ever one, why wrap it in an array of length 1?